core/kernel/dt.c

```diff
@@ -472,6 +496,203 @@ void *get_secure_dt(void)
+} parent_node_cache;
+
+static int cmp_parent_node_cache(const void *a, const void *b)
+{
+	const struct parent_node_cache *cell_a = a;
+	const struct parent_node_cache *cell_b = b;
+
+	return CMP_TRILEAN(cell_a->node_offset, cell_b->node_offset);
+}
+
+static struct parent_node_cache *find_cached_parent_node(const void *fdt,
+							 int node_offset)
+{
+	struct parent_node_cache target = {
+		.node_offset = node_offset,
+	};
+
+	if (!cached_node_info_fdt || fdt != cached_node_info_fdt)
+		return NULL;
+
+	return bisect_equal(parent_node_cache.array, parent_node_cache.count,
+			    sizeof(target), &target, cmp_parent_node_cache);
+}
+
+int fdt_find_cached_parent_node(const void *fdt, int node_offset,
+				int *parent_offset)
+{
+	struct parent_node_cache *cell = NULL;
+
+	cell = find_cached_parent_node(fdt, node_offset);
+	if (!cell)
+		return -FDT_ERR_NOTFOUND;
+
+	*parent_offset = cell->parent_offset;
+
+	return 0;
+}
+
+int fdt_find_cached_parent_reg_cells(const void *fdt, int node_offset,
+				     int *address_cells, int *size_cells)
+{
+	struct parent_node_cache *cell = NULL;
+
+	cell = find_cached_parent_node(fdt, node_offset);
+	if (!cell)
+		return -FDT_ERR_NOTFOUND;
+
+	if (address_cells)
+		*address_cells = cell->address_cells;
+
+	if (size_cells)
+		*size_cells = cell->size_cells;
+
+	return 0;
+}
+
+static void release_parent_node_cache(void)
+{
+	free(parent_node_cache.array);
+	parent_node_cache.array = NULL;
+	parent_node_cache.count = 0;
+	parent_node_cache.alloced_count = 0;
+}
+
+static TEE_Result enlarge_parent_node_cache(void)
+{
+	if (parent_node_cache.count + 1 > parent_node_cache.alloced_count) {
+		/* Allocate by chunk of 64 cells for efficiency */
+		size_t new_count = parent_node_cache.alloced_count + 64;
+		struct parent_node_cache *new = NULL;
+
+		new = realloc(parent_node_cache.array,
+			      sizeof(*parent_node_cache.array) * new_count);
+		if (!new)
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		parent_node_cache.array = new;
+		parent_node_cache.alloced_count = new_count;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result add_parent_node_info(const void *fdt __maybe_unused,
+				       int parent_offset, int node_offset,
+				       int address_cells, int size_cells)
+{
+	size_t count = parent_node_cache.count;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = enlarge_parent_node_cache();
+	if (res)
+		return res;
+
+	parent_node_cache.array[count] = (struct parent_node_cache){
+		.node_offset = node_offset,
+		.parent_offset = parent_offset,
+		.address_cells = address_cells,
+		.size_cells = size_cells,
+	};
+
+	parent_node_cache.count++;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result add_node_cached_info(const void *fdt, int node_offset)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
+	int subnode_offset = 0;
+	size_t addr_cells = 0;
+	size_t size_cells = 0;
+
+	cuint = fdt_getprop(fdt, node_offset, "#address-cells", NULL);
+	if (cuint)
+		addr_cells = (int)fdt32_to_cpu(*cuint);
```

Comment:
I see that libfdt does not go 1 parent-level up to get these information. It would make sense but I don't think this is how DT support is implemented.